### PR TITLE
feat(mint-client): include federation join info in `OOBNotes`

### LINKED ISSUE #4048: E-cash should optionally contain invite code
That way receiving e-cash from unknown federations becomes possible. While this isn't recommended there seems to be strong user demand.

### PR DESCRIPTION
**Pull request by elsirion**
Fixes #4048

### CODE REVIEW

File: modules/fedimint-mint-client/src/lib.rs [hunk 1]
```diff
@@ -110,11 +119,25 @@ impl OOBNotes {
         ])
     }
 
+    pub fn new_with_invite(notes: TieredMulti<SpendableNote>, invite: InviteCode) -> Self {
+        Self(vec![
+            // FIXME: once we can break compatibility with 0.2 we can remove the prefix in case an
```
Comment (Maan2003) [inline modules/fedimint-mint-client/src/lib.rs:124]: maybe `FIXME(0.2)` so we can grep for it later

File: modules/fedimint-mint-client/src/lib.rs [hunk 2]
```diff
@@ -92,6 +93,14 @@ pub struct OOBNotes(Vec<OOBNotesData>);
 enum OOBNotesData {
```
Comment (Maan2003) [inline modules/fedimint-mint-client/src/lib.rs:93]: OOBNotesData was a good idea!

File: fedimint-client/src/module/mod.rs [hunk 3]
```diff
@@ -434,6 +434,22 @@ where
...
+        let (any_guardian_id, any_guardian_url) = cfg
+            .api_endpoints
+            .iter()
+            .next()
+            .expect("A federation always has at least one guardian");
+        InviteCode::new(
+            any_guardian_url.url.clone(),
+            *any_guardian_id,
+            cfg.federation_id(),
+        )
```
Comment (Maan2003) [inline fedimint-client/src/module/mod.rs:450]: would be nice to also have this function on config type

File: modules/fedimint-mint-client/src/lib.rs [hunk 4]
```diff
@@ -92,6 +93,14 @@ pub struct OOBNotes(Vec<OOBNotesData>);
...
     Notes(TieredMulti<SpendableNote>),
     FederationIdPrefix(FederationIdPrefix),
+    /// Invite code to join the federation by which the e-cash was issued
+    ///
+    /// Introduce in 0.3.0
+    Invite {
```
Comment (dpc) [inline modules/fedimint-mint-client/src/lib.rs:99]: If this is present the `FederationIdPrefix` is redundant, though we might start to omit it in some future. Though we might already change code that uses `FederationIdPrefix` to use see if `Invite` is available.

Comment (elsirion) [reply modules/fedimint-mint-client/src/lib.rs:99]: Is this a documentation request?

Comment (dpc) [reply modules/fedimint-mint-client/src/lib.rs:99]: No. Code request, unless I'm missing something.

File: modules/fedimint-mint-client/src/lib.rs [hunk 5]
```diff
@@ -110,11 +119,25 @@ impl OOBNotes {
...
+            },
+        ])
+    }
+
     pub fn federation_id_prefix(&self) -> FederationIdPrefix {
         self.0
             .iter()
             .find_map(|data| match data {
                 OOBNotesData::FederationIdPrefix(prefix) => Some(*prefix),
+                OOBNotesData::Invite { federation_id, .. } => Some(federation_id.to_prefix()),
```
Comment (elsirion) [inline modules/fedimint-mint-client/src/lib.rs:140]: @dpc feels like this is what you were asking for in your comment?

File: modules/fedimint-mint-client/src/lib.rs (hunk 4 above)
Comment (elsirion) [reply modules/fedimint-mint-client/src/lib.rs:99]: https://github.com/fedimint/fedimint/pull/4231#discussion_r1477646687

File: modules/fedimint-mint-client/src/lib.rs (hunk 5 above)
Comment (dpc) [reply modules/fedimint-mint-client/src/lib.rs:140]: :+1:

**Review by dpc — APPROVED** [review]

**Review by Maan2003 — APPROVED** [review]

**Pull request merged**